include/vast/ABI/MLIRTypeInfo.hpp

```diff
@@ -0,0 +1,252 @@
+            : dl(dl), module_op(module_op)
+        {}
+
+        // Rough draft of the interface (or concept) to implement by any IR that wants
+        // to use the ABI classification.
+        static bool is_void(mlir_type t);
+        static bool is_complex(mlir_type t);
+
+        static bool is_scalar_integer(mlir_type t);
+        static bool is_scalar_float(mlir_type t);
+
+        static bool is_scalar(mlir_type t);
+        static bool is_aggregate(mlir_type t);
+        static bool is_record(mlir_type t);
+        static bool is_struct(mlir_type t);
+        static bool is_array(mlir_type t);
+
+        static bool can_be_passed_in_regs(mlir_type t);
+        static bool has_unaligned_field(mlir_type t);
+
+        // Pointers, references etc
+        static bool represents_pointer(mlir_type t);
+        static bool is_pointer(mlir_type t);
+
+        static bool can_be_promoted(mlir_type t);
+        static mlir_type prepare(mlir_type t);
+
+        // TODO: Move into different abstraction, pointer size is dependent on underlying
+        //       cpu architecture not the IR.
+        static std::size_t pointer_size();
+
+        std::size_t size(mlir_type t);
+        std::size_t align(mlir_type t);
+
+        mlir_type iN(std::size_t s);
+        mlir_type fN(std::size_t s);
+
+        static mlir_type as_pointer(mlir_type t);
+
+        mlir_type type_at_offset(mlir_type t, std::size_t offset, const auto &accept);
+
+        auto field_containing_offset(mlir_type t, std::size_t offset)
+            -> std::optional< std::tuple< mlir_type, std::size_t > >;
+
+        gap::generator< mlir_type > mock_array_fields(hl::ArrayType array_type);
+
+        auto fields(mlir_type type);
+
+        using array_info_t = std::tuple< std::optional< std::size_t >, mlir_type >;
+        array_info_t array_info(mlir_type t);
+    };
+
+    bool vast_type_info::is_void(mlir_type t) {
+        return mlir::isa< mlir::NoneType >(t);
+    }
+
+    bool vast_type_info::is_scalar(mlir_type t) {
+        // TODO(abi): Also complex is an option in ABI.
+        return !is_aggregate(t);
+    }
+
+    bool vast_type_info::is_complex(mlir_type t) {
+        VAST_UNREACHABLE("Missing support for complex types!");
+    }
+
+    // TODO(abi): `SubElementTypeInterface` is not good enough, since for
+    //             example hl::PointerType implements it.
+    // TODO(abi): Figure how to better handle this than manual listing.
+    bool vast_type_info::is_aggregate(mlir_type t) {
+        return mlir::isa< hl::RecordType, hl::ElaboratedType, hl::ArrayType >(t);
+    }
+
+    bool vast_type_info::is_record(mlir_type t) {
+        if (t.isa< hl::RecordType >()) {
+            return true;
+        }
+        if (auto et = t.dyn_cast< hl::ElaboratedType >()) {
+            return is_record(et.getElementType());
+        }
+        return false;
+    }
+
+    bool vast_type_info::is_struct(mlir_type t) {
+        // TODO(abi): Are these equivalent?
```

Comment:
yes keep only one

Reply:
I will adjust/rework the comment - as this follows the clang codebase this will be expanded once classes are in the picture.